[EmbedUrlProcessorServlet] Non-deterministic behavior of Class.getDeclaredFields() might fail test EmbedUrlProcessorServletTest.testUrlWithRegisteredProvider()

<!-- 

Before sending a bug report please check if the latest release (if you're not already using it) fixes your problem. Alternatively, you could
install the latest snapshot version from the master branch.
 -->

## Bug Report

**Current Behavior**
- `EmbedUrlProcessorServlet.doGet()` writes the result fetch into `UrlProcessor.Result` as `JSON` using `ObjectMapper` at https://github.com/adobe/aem-core-wcm-components/blob/3524e33424bf6831faca014284c5bad803d3744d/bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/embed/EmbedUrlProcessorServlet.java#L69-L84
- However, `ObjectMapper.writeValue()` converts the value into `JSON` by fetching the fields of `UrlProcessorResultImpl`, which are `processor` and `options`, using Java reflection `Class.getDeclaredFields()` at `AnnotatedFieldCollector.java:73`
- The official [Javadoc](https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#getDeclaredFields--) of `Class.getDeclaredFields()` mentions that the elements are not returned in any particular order. This causes issue where the order of class fields are expected to be in a specific order. 
- Specifically, this assumption could be observed in `com.adobe.cq.wcm.core.components.internal.servlets.embed.EmbedUrlProcessorServletTest.testUrlWithRegisteredProvider()`, where the JSON string result from the `EmbedUrlProcessorServlet.doGet()` is expected to be in the order of `process` followed by `options` at https://github.com/adobe/aem-core-wcm-components/blob/3524e33424bf6831faca014284c5bad803d3744d/bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/servlets/embed/EmbedUrlProcessorServletTest.java#L74
- But this order might be changed based on the order of fields returned by `Class.getDeclaredFields()`, which could result in the following error
```
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   EmbedUrlProcessorServletTest.testUrlWithRegisteredProvider:78 Does not match the expected response output. ==> expected: <{"processor":"pinterest","options":{"pinId":"99360735500167749"}}> but was: <{"options":{"pinId":"99360735500167749"},"processor":"pinterest"}>
[INFO] 
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0
[INFO] 
```

#### Steps to reproduce
- We could observe the above mentioned issue by running the test through [NonDex](https://github.com/TestingResearchIllinois/NonDex), which shuffles the order of elements returned by `Class.getDeclaredFields()`
```
# Clone the Repo
https://github.com/adobe/aem-core-wcm-components.git

# Compile the module byte-buddy-dep
mvn clean install -pl bundles/core -am -DskipTests

# (Optional) Run the unit test
mvn -pl bundles/core test -Dtest=com.adobe.cq.wcm.core.components.internal.servlets.embed.EmbedUrlProcessorServletTest#testUrlWithRegisteredProvider

# Run the unit test using NonDex
mvn -pl bundles/core edu.illinois:nondex-maven-plugin:2.1.1:nondex -Dtest=com.adobe.cq.wcm.core.components.internal.servlets.embed.EmbedUrlProcessorServletTest#testUrlWithRegisteredProvider
```

**Expected behavior/code**
- For any amount of runs, the test should have passes successfully (with `NonDex` too). 
- The expected order of elements might not necessarily be `processor` followed by `options`. 

**Environment**
- Core Components version: `2.23.5-SNAPSHOT`
- Java version: `Java(TM) SE Runtime Environment (build 1.8.0_381-b09)`

**Possible Solution 1**
- Since the test validates the data of the json and not the order, we can convert the strings to `JsonNode` and compare the nodes.
```
+/-  com.adobe.cq.wcm.core.components.internal.servlets.embed.EmbedUrlProcessorServletTest.testUrlWithRegisteredProvider()

-    String expectedOutput = "{\"processor\":\"pinterest\",\"options\":{\"pinId\":\"99360735500167749\"}}";
-    assertEquals(expectedOutput, context.response().getOutputAsString(), "Does not match the expected response output.");

+    ObjectMapper mapper = new ObjectMapper();
+    JsonNode expected = mapper.readTree("{\"processor\":\"pinterest\",\"options\":{\"pinId\":\"99360735500167749\"}}");
+    JsonNode actual = mapper.readTree(context.response().getOutputAsString());
+    assertEquals(expected, actual, "Does not match the expected response output.");
```
**Possible Solution 2**
- Since the order of elements might change inside `ObjectMapper` and there are only two possible strings returned, we could expect both the possibilities in the test.
```
+/-  com.adobe.cq.wcm.core.components.internal.servlets.embed.EmbedUrlProcessorServletTest.testUrlWithRegisteredProvider()

-    String expectedOutput = "{\"processor\":\"pinterest\",\"options\":{\"pinId\":\"99360735500167749\"}}";
-    assertEquals(expectedOutput, context.response().getOutputAsString(), "Does not match the expected response output.");

+    String expectedOutput1 = "{\"processor\":\"pinterest\",\"options\":{\"pinId\":\"99360735500167749\"}}";
+    String expectedOutput2 = "{\"options\":{\"pinId\":\"99360735500167749\"},\"processor\":\"pinterest\"}";
+    String actualOutput = context.response().getOutputAsString();
+    assertTrue(expectedOutput1.equals(actualOutput) || expectedOutput2.equals(actualOutput), "Does not match the expected response output.");
```
- **If the proposed solution looks good, I can create a PR.**

--- a/bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/servlets/embed/EmbedUrlProcessorServletTest.java
+++ b/bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/servlets/embed/EmbedUrlProcessorServletTest.java
@@ -32,6 +32,7 @@ import io.wcm.testing.mock.aem.junit5.AemContext;
 import io.wcm.testing.mock.aem.junit5.AemContextExtension;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 
 @ExtendWith(AemContextExtension.class)
 public class EmbedUrlProcessorServletTest {
@@ -71,9 +72,11 @@ public class EmbedUrlProcessorServletTest {
     public void testUrlWithRegisteredProvider() throws Exception {
         context.request().setQueryString("url=https://www.pinterest.com/pin/99360735500167749/");
         servlet.doGet(context.request(), context.response());
-        String expectedOutput = "{\"processor\":\"pinterest\",\"options\":{\"pinId\":\"99360735500167749\"}}";
         assertEquals(HttpServletResponse.SC_OK, context.response().getStatus(), "Expected the 200 status code.");
         assertEquals("application/json;charset=utf-8", context.response().getContentType(), "Expected the JSON content type.");
-        assertEquals(expectedOutput, context.response().getOutputAsString(), "Does not match the expected response output.");
+        String expectedOutput1 = "{\"processor\":\"pinterest\",\"options\":{\"pinId\":\"99360735500167749\"}}";
+        String expectedOutput2 = "{\"options\":{\"pinId\":\"99360735500167749\"},\"processor\":\"pinterest\"}";
+        String actualOutput = context.response().getOutputAsString();
+        assertTrue(expectedOutput1.equals(actualOutput) || expectedOutput2.equals(actualOutput), "Does not match the expected response output.");
     }
 }

--- a/bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/servlets/embed/EmbedUrlProcessorServletTest.java
+++ b/bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/servlets/embed/EmbedUrlProcessorServletTest.java
@@ -74,10 +74,11 @@ public class EmbedUrlProcessorServletTest {
     public void testUrlWithRegisteredProvider() throws Exception {
         context.request().setQueryString("url=https://www.pinterest.com/pin/99360735500167749/");
         servlet.doGet(context.request(), context.response());
+        String expectedOutput = "{\"processor\":\"pinterest\",\"options\":{\"pinId\":\"99360735500167749\"}}";
         assertEquals(HttpServletResponse.SC_OK, context.response().getStatus(), "Expected the 200 status code.");
         assertEquals("application/json;charset=utf-8", context.response().getContentType(), "Expected the JSON content type.");
         ObjectMapper mapper = new ObjectMapper();
-        JsonNode expected = mapper.readTree("{\"processor\":\"pinterest\",\"options\":{\"pinId\":\"99360735500167749\"}}");
+        JsonNode expected = mapper.readTree(expectedOutput);
         JsonNode actual = mapper.readTree(context.response().getOutputAsString());
         assertEquals(expected, actual, "Does not match the expected response output.");
     }

--- a/bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/servlets/embed/EmbedUrlProcessorServletTest.java
+++ b/bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/servlets/embed/EmbedUrlProcessorServletTest.java
@@ -28,11 +28,13 @@ import com.adobe.cq.wcm.core.components.context.CoreComponentTestContext;
 import com.adobe.cq.wcm.core.components.internal.services.embed.PinterestUrlProcessor;
 import com.adobe.cq.wcm.core.components.services.embed.UrlProcessor;
 import com.adobe.cq.wcm.core.components.testing.Utils;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
 import io.wcm.testing.mock.aem.junit5.AemContext;
 import io.wcm.testing.mock.aem.junit5.AemContextExtension;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertTrue;
 
 @ExtendWith(AemContextExtension.class)
 public class EmbedUrlProcessorServletTest {
@@ -74,9 +76,9 @@ public class EmbedUrlProcessorServletTest {
         servlet.doGet(context.request(), context.response());
         assertEquals(HttpServletResponse.SC_OK, context.response().getStatus(), "Expected the 200 status code.");
         assertEquals("application/json;charset=utf-8", context.response().getContentType(), "Expected the JSON content type.");
-        String expectedOutput1 = "{\"processor\":\"pinterest\",\"options\":{\"pinId\":\"99360735500167749\"}}";
-        String expectedOutput2 = "{\"options\":{\"pinId\":\"99360735500167749\"},\"processor\":\"pinterest\"}";
-        String actualOutput = context.response().getOutputAsString();
-        assertTrue(expectedOutput1.equals(actualOutput) || expectedOutput2.equals(actualOutput), "Does not match the expected response output.");
+        ObjectMapper mapper = new ObjectMapper();
+        JsonNode expected = mapper.readTree("{\"processor\":\"pinterest\",\"options\":{\"pinId\":\"99360735500167749\"}}");
+        JsonNode actual = mapper.readTree(context.response().getOutputAsString());
+        assertEquals(expected, actual, "Does not match the expected response output.");
     }
 }

